src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/AesGcm.Unix.cs

```diff
@@ -0,0 +1,202 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading.Tasks;
+using Internal.Cryptography;
+using Microsoft.Win32.SafeHandles;
+
+namespace System.Security.Cryptography
+{
+    partial class AesGcm
+    {
+        // See: aes_gcm_ctrl: conditions related to EVP_CTRL_GCM_SET_TAG/EVP_CTRL_GCM_GET_TAG
+        public static KeySizes TagByteSizes { get; } = new KeySizes(1, 16, 1);
+        private SafeEvpCipherCtxHandle _ctxHandle;
+
+        public AesGcm(ReadOnlySpan<byte> key)
+        {
+            if (key == null)
+                throw new ArgumentNullException(nameof(key));
+
+            int keyLength = key.Length * 8;
+            AesAEAD.CheckKeySize(keyLength);
+
+            _ctxHandle = Interop.Crypto.EvpCipherCreatePartial(
+                        GetCipher(keyLength),
+                        keyLength,
+                        0);
+
+            Interop.Crypto.CheckValidOpenSslHandle(_ctxHandle);
+
+            ref byte nullRef = ref MemoryMarshal.GetReference(Span<byte>.Empty);
+            if (!Interop.Crypto.EvpCipherSetKeyAndIV(_ctxHandle, ref MemoryMarshal.GetReference(key), ref nullRef, -1))
+            {
+                throw Interop.Crypto.CreateOpenSslCryptographicException();
+            }
+
+            Debug.Assert(NonceByteSizes.MinSize == 12 && NonceByteSizes.MaxSize == 12);
+            if (!Interop.Crypto.EvpAesGcmSetNonceLength(_ctxHandle, 12))
+            {
+                throw Interop.Crypto.CreateOpenSslCryptographicException();
+            }
+        }
+
+        public void Encrypt(ReadOnlySpan<byte> nonce, ReadOnlySpan<byte> plaintext, Span<byte> ciphertext, Span<byte> tag, ReadOnlySpan<byte> associatedData = default)
+        {
+            AesAEAD.CheckArguments(nonce, plaintext, ciphertext, tag);
+            CheckParameters(nonce.Length, tag.Length);
+
+            ref byte nullRef = ref MemoryMarshal.GetReference(Span<byte>.Empty);
+
+            if (!Interop.Crypto.EvpCipherSetKeyAndIV(_ctxHandle, ref nullRef, ref MemoryMarshal.GetReference(nonce), 1 /* encrypting */))
+            {
+                throw Interop.Crypto.CreateOpenSslCryptographicException();
+            }
+
+            if (associatedData.Length != 0)
+            {
+                if (!Interop.Crypto.EvpCipherUpdate(_ctxHandle, ref nullRef, out _, ref MemoryMarshal.GetReference(associatedData), associatedData.Length))
```

Comment:
If the rule is that it is AD when the output buffer is null, a comment would be good.

// Set the associated data by calling update with a null output